Multiple Renders Fix

fordacious opened the pull request:
Decoupled webgl properties (buffers, initialization etc...) of renderable objects attributes into a separate object. This allows you to render objects (e.g scenes, meshes and materials) between different renderers without having them tied to the first renderer they are rendered to.

Examples are prefixed with "webgl_multiple_renderers_". I also have other examples which I have not committed.

Commit:
decoupled webgl properties and buffers from renderable objects and attributes. Scenes and objects no longer coupled to renderers and may be used between multiple renderers.

Commit:
removed comment

Commit:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

Commit:
added multiple renderers examples

Dynasty9 commented:
:+1: to this

dubejf commented on src/renderers/WebGLRenderer.js:
```diff
@@ -774,16 +779,22 @@ THREE.WebGLRenderer = function ( parameters ) {
+
+		}
+
+		if ( object.hasPositions && ! objectRendererWebGLProps[object.uuid].__webglVertexBuffer ) objectRendererWebGLProps[object.uuid].__webglVertexBuffer = _gl.createBuffer();
```
You should store the value of `objectRendererWebGLProps[...uuid]` in a var to avoid repeated lookup.

fordacious replied:
Fixed

Commit:
added temporary variable to lower lookups

Commit:
added temporary var to lower lookup

WestLangley commented:
You probably only need one example -- the horse example is sufficient. But even that example can be implemented with a single renderer and multiple viewports.

What is _your_ use case for which multiple renderers is required?

Check the console for errors/warnings.

mrdoob commented:
> What is your use case for which multiple renderers is required?

One use case is to be able to render the same scene twice, with and without antialias.

I think this is great! We were heading this way. This cleans things a little bit.

dubejf commented on src/renderers/WebGLRenderer.js:
```diff
@@ -653,24 +658,24 @@ THREE.WebGLRenderer = function ( parameters ) {
+			_gl.deleteTexture( objectRendererWebGLProps[texture.uuid].__image__webglTextureCube );
 
-			delete texture.image.__webglTextureCube;
+			delete objectRendererWebGLProps[texture.uuid].__image__webglTextureCube;
```
Should you also call `delete objectRendererWebGLProps[texture.uuid]` ?

fordacious replied:
Changed for all deallocations.

dubejf commented on src/renderers/WebGLRenderer.js:
```diff
@@ -653,24 +658,24 @@ THREE.WebGLRenderer = function ( parameters ) {
 	var deallocateTexture = function ( texture ) {
 
-		if ( texture.image && texture.image.__webglTextureCube ) {
+		if ( texture.image && objectRendererWebGLProps[texture.uuid].__image__webglTextureCube ) {
```
There are a lot of unconditional access to `objectRendererWebGLProps[....uuid].prop`.

Could `objectRendererWebGLProps[....uuid]` ever return undefined?

fordacious replied:
Added checks at the beginning of functions in which the webgl properties are accessed.

Another strategy would be to encapsulate access of webgl properties with a getter and setter, (like what's going on in webglObject with attribute buffers).

It would looks something like"

```javascript

function getWebGLProps (obj) {
   if ( ! objectRendererWebGLProps[object.uuid]) {
    
        objectRendererWebGLProps[object.uuid] = {};

   }

   return objectRendererWebGLProps[object.uuid];
}

getWebGLProps(texture)
```

Seems better design, but I don't think the extra condition / lookup is worth it.

mrdoob replied:
> It would looks something like"

> ```javascript
>
> function getWebGLProps (obj) {
>    if ( ! objectRendererWebGLProps[object.uuid]) {
>     
>        objectRendererWebGLProps[object.uuid] = {};
>
>   }
>
>    return objectRendererWebGLProps[object.uuid];
> }

> getWebGLProps(texture)
```

I actually prefer that design. It aligns more with how I've been doing [WebGLObjects](https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLObjects.js), [WebGLGeometries](https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLGeometries.js) and so on.

Maybe you could move all this into a new class named `WebGLProps`?

fordacious replied:
Added the class and moved all the checks into it.

WestLangley commented:
>What is _your_ use case for which multiple renderers is required?
>>One use case is to be able to render the same scene twice, with and without antialias.

Would someone want to do that?

I was hopeful there was a useful case that could be demonstrated in an example -- a case in which two renderers is _required_.

Commit:
webgl props are removed in one go in object deallocation, added better example, removed other examples, added check for properties at entrance of each function that uses properties

Commit:
fixed examples names

Commit:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

Conflicts:
	src/renderers/WebGLRenderer.js

fordacious commented:
I added an example in which a scene is rendered to multiple canvases which move dynamically on the page. A lot of the use cases for multiple renderers can be achieved without the use of multiple renderers, however there are cases in which it is a lot simpler and cleaner to use multiple renderers which share objects or scenes. 

My case specifically is that I preload a bunch of meshes exported from blender, and share them dynamically between moving canvasses (each rendered on separate renderers and scenes) and it is far simpler to manage with this fix. If you would like to see an example resembling this, I can put one together.

I also believe that decoupling webgl objects and state from the renderable objects themselves is a good idea in general as it helps keep complexity down and makes the library flexible.

Commit:
cosmetic example changes

WestLangley commented:
>  If you would like to see an example resembling this, I can put one together.

I think a single example showing the use of multiple renderers is sufficient. Which example is up to you. : - )

Commit:
removed cube example

fordacious commented:
Cool, I removed the cube example. Only one now is the canvases one.

fordacious commented:
@mrdoob anything that needs to be done?

Commit:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

Commit:
Added class WebGLProperties, webgl properties now stored in that class. Checks for uuids and existance of properties encapsulated in this class

mrdoob commented:
Much nicer!

mrdoob commented:
>>> What is _your_ use case for which multiple renderers is required?

>> One use case is to be able to render the same scene twice, with and without antialias.

> Would someone want to do that?

#189 #4221 #5987 #6062 ...

This is a long standing issue. Other renderers have no issues when rendering the same scene. It was a problem with how `WebGLRenderer` was implemented. It polutes the scenegraph with buffers instead of storing the buffers internaly (which is what this PR does). Not only this allows to be able to render the same scene from different `WebGLRenderer`s but it also allows to restore the scene if webgl loses its context.

Another use case is that three.js editor breaks if you change the renderer to another renderer and back to `WebGLRenderer` because the new renderer is finding buffers created even if they're from a different context/renderer.

It also makes the renderer much more tidier and follows what I've been doing the last months. Still needs some refactoring but it's a step in the right direction 😊

WestLangley commented:
@mrdoob Yes, yes. I support the code improvement. : - )

I was asking a use-case question: What is the use case for which two renderers are required -- a use case that cannot be implemented with a single renderer. I was requesting an example that demonstrated that use case.

Somehow, I feel that I am getting an answer to a different question...

mrdoob commented:
> I was asking a use-case question: What is the use case for which two renderers are required -- a use case that cannot be implemented with a single renderer. I was requesting an example that demonstrated that use case.

Ah! I see what you mean. Yes, probably a example is not really needed as it's actually bad practice.

Mmh, maybe we could have an example using `WebGLRenderer` and `WebGLDeferredRenderer`?

WestLangley commented:
> Yes, probably a example is not really needed as it's actually bad practice.

That is what I thought. Then let's not encourage it.

... unless someone can provide a compelling use case.

mrdoob commented:
Indeed. However, we could have a `misc_ubiquity_test3` using `WebGLRenderer` and `WebGLDeferredRenderer`.

WestLangley commented:
Perhaps a use case would be a web page with two webgl canvases -- each rendering the same scene, but with, say, a different view?

That would require two renderers, no?

mrdoob commented:
Nah, you can (and should) do that with viewports. In fact, check out `webgl_multiple_elements.html`. All these elements are just one `WebGLRenderer`.

WestLangley commented:
No, in the use case I mentioned, I would want separate canvases  -- the canvases are not adjacent on the webpage.

gero3 commented:
@WestLangley,
2 use cases I heard of are:  
- blog post showing muliple steps of a 3d scene.
- To restore a crashed webgl canvas, it also needs to remove all cached objects.

tonnot commented:
@WestLangley  @mrdoob  . Yes there are things to be done using viewport's but I have  divs, dialog and/or independent windows in which I would  want to have different shows of the same scene / scenes

Commit:
removed adding of uuid on addAttribute

Commit:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

Commit:
renamed webglproperties method names. Now correctly accounts for interleaved attributes sharing buffers

Commit:
correctly merged from dev

Commit:
merge

fordacious commented:
I changed the method names, added uuids to the necessary classes and added a check for interleavedAttributes in webGLObjects (to account for buffers shared between attributes).

Would you like me to delete all webGLProperties on context loss?

I also accidentally removed the compiled builds and committed that. Will recommit with latest dev builds.

Commit:
added dev build files

WestLangley commented:
No need to include build files in pull requests.

See the Wiki article [How to Contribute to three.js](https://github.com/mrdoob/three.js/wiki/How-to-contribute-to-three.js).

mrdoob commented:
> Would you like me to delete all webGLProperties on context loss?

That would be great! 😃

gocoffeecup commented:
I'm looking forward to this PR.

@WestLangley as usecase consider a hud which is composed with html in which you would like to render different meshs, you need another canvas to overlay the hud and the main scene.

WestLangley commented:
@gocoffeecup HUDs can be implemented via viewports and a single renderer. http://threejs.org/examples/webgl_shadowmap.html.

Granted, you may want a separate, overlaid, draggable div that renders the same scene, say, from a different view. I agree, that is a compelling use case for this PR.

dubejf commented:
Is this going to cause a problem with `material.needsUpdate` and `texture.needsUpdate` ? First renderer resets flag to `false`, second renderer doesn't see the update... ?

If that's the case, maybe the flag `needsUpdate` can be changed to setter property. Assigning `needsUpdate` would increment a counter.

```
set needsUpdate( value ) {

	this.generationCount ++;

},
```


In the renderer, instead of checking for `if ( texture.needsUpdate === true ) { ... update ... } `

```
var textureWebGLProperties = webGLProps.get( texture );

if ( texture.generationCount !== textureWebGLProperties.generationCount ) {

	... update ...

}

```

It's not a regression because multiple renderers are not currently supported. For clarity, this can probably be implemented in another patch.

benaadams commented:
> If that's the case, maybe the flag needsUpdate can be changed to setter property. Assigning needsUpdate would increment a counter.

That could cause a problem if you are doing dispatched updates or dirty rectangle updates as it would be marked multiple times? Would be easier to remark between dual renders than do something automatic else would be optimising for a bad practice edge-case and not general use?

dubejf commented:
@benaadams I'm not sure that I follow you. If you set `texture.needsUpdate = true` 20 times in-between frames, the generation counter for the texture will increase by 20, but the renderer will only care that the cached WebGLProp is not at the latest version. It will sync the buffers once and won't need to be updated again until the next `needsUpdate`. Same behavior has before.

Similarly, if you change the data of a `THREE.Texture` but do not wish to publish those change to the renderers yet, simply omit `texture.needsUpdate`.

I edited the example with more comments.

benaadams commented:
> If you set texture.needsUpdate = true 20 times in-between frames, the generation counter for the texture will increase by 20, but the renderer will only care that the cached WebGLProp is not at the latest version.

Ah gotcha, version counter; works well :)

benaadams commented:
In related news this thread on the Public WebGL list might be worth paying attention to: https://www.khronos.org/webgl/public-mailing-list/archives/1506/msg00049.html

Commit:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

Commit:
delete all webglProperties on loss of context

fordacious commented:
Now deletes all webglProperties stored by the renderer and its webglObjects on context loss.

@dubejf That's a good idea. It would need to be implemented on other objects as well, such as attributes and cubemaps. Currently you have to manually set needsUpdate before rendering on each renderer. Your solution looks like it would solve this. Might be too big for this PR though so another PR might be in order. On this topic, is there a particular reason as to why needsUpdate needs to be manually called as opposed to being set on object changes (e.g. performance / more control)?

fordacious commented:
@mrdoob all good?

mrdoob commented:
Looks good! Thanks! 😊

mlepage commented:
Hello, my use case at the moment is prototyping some image processing algorithms. I want to process the pixels of an image (loaded into a texture) into separate outputs, so I can do further passes on them, and then later recombine them in a final pass. (E.g. say I wanted the first pass to split out the high and low frequency components of the image, then do something with them in a few passes, then finally recombine them). It seems like that requires multiple renderer targets, does that sound right? Thanks for all this work, we really do appreciate it!

mrdoob commented:
@fordacious seems like this PR broke the example [webgl_materials_cubemap](https://cdn.rawgit.com/mrdoob/three.js/ce6c34d94359c712800afee53dc4614d13937760/examples/webgl_materials_cubemap.html) ([before](https://cdn.rawgit.com/mrdoob/three.js/9760cf7af778a627a2af46ad97dbe1921cb32f69/examples/webgl_materials_cubemap.html)). Do you mind taking a look?

WestLangley commented:
It's a `needsUpdate` problem. After entering 

    scene.children[ 3 ].material.envMap.needsUpdate = true;

into the console, the scene renders correctly.

mrdoob commented:
Hmm... True. The cubemap is also upside down too. If I set `refractionRatio` to `1` you can see it clearly.

![screen shot 2015-07-12 at 00 35 02](https://cloud.githubusercontent.com/assets/97088/8636421/e3ef8baa-282d-11e5-8f2d-37684ed5b1ec.png)

WestLangley commented:
To clarify, the cube map is only upside down for refraction. It is correct for reflection.

This appears to be related to the problem:

	var refractionCube = new THREE.CubeTexture( reflectionCube.image, THREE.CubeRefractionMapping );
	refractionCube.format = THREE.RGBFormat;

This works:

	var refractionCube = THREE.ImageUtils.loadTextureCube( urls, THREE.CubeRefractionMapping );
	refractionCube.format = THREE.RGBFormat;

fordacious commented:
Confirmed its broken on my dev branch. Will look into it.

WestLangley commented:
`THREE.ImageUtils.loadTextureCube()` contains this code: 

		// no flipping needed for cube textures
		texture.flipY = false;

That code should be moved into the `CubeTexture` constructor. Then, I expect everything will work as expected.

This change should be in a separate PR, as it impacts other files.

We also need to decide if it is a `CubeTexture` or a `TextureCube`.

fordacious commented:
Moving `texture.flipY = false` to the CubeTexture class solves that issue.

The blank texture problem is caused by the texture having needsUpdate set to false and thus the texture is never created (and textureProperties.__image__webglTextureCube is undefined). This may be solved by changing the `setCubeTexture` function in WebGLRenderer from

```javascript
if ( texture.needsUpdate ) {
	if ( ! textureProperties.__image__webglTextureCube ) {

		...

	}

	...
```

to

```javascript
if ( ! textureProperties.__image__webglTextureCube ) {
	...
	texture.needsUpdate = true;
}

if ( texture.needsUpdate ) {
	...
```
this will ensure the texture is initialized and updated properly. 

Haven't checked all examples yet but it seems solid. Want me to create a PR with these changes?

WestLangley commented:
> Want me to create a PR with these changes?

@fordacious Separate PRs would be preferable, if you don't mind. Thanks.

mrdoob commented:
Yes please!

fordacious commented:
pr's
#6839 
#6838

DreamingSandman commented:
Thanks! This will help a lot.
I'm working with gazebo's web client gzweb (HumanBrainProject) and this is exactly what we needed. Multiple adjustable divs/canvases where we render the same scene with different effects and views (e.g. parallel user view, robot camera, etc.).
Had to use a workaround for now but this will make things a lot cleaner.

mrdoob commented:
@DreamingSandman You should consider using viewports instead if you can though!

plmrry commented:
> You should consider using viewports instead if you can though!

@mrdoob I believe you, but could you explain why, exactly? 

I assume there is some major duplication of resources going on when you use two renderers, but I'm wondering to what extent this will actually affect performance. For simple scenes, maybe it wouldn't matter?

mrdoob commented:
It's basically 2x GPU memory.

plmrry commented:
Ah, makes sense. Thanks.

DreamingSandman commented:
What's preventing the viewport based approach for me is shadowmaps in combination with scissor test (r71): https://github.com/mrdoob/three.js/issues/3532. Would the changes from r72 help with that?

mrdoob commented:
@DreamingSandman: Fixed in `r74dev` 😊 faebb42ee1146d604f8c6c7e3ed6c85dc78b2f4b

DreamingSandman commented:
Awesome :D.
